Update linked product search to use GB combobox in order to fix timing issue

louwie17 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR updates the `ProductSelect` component used in the linked list search to use the Combobox control from Gutenberg instead of our internal SelectControl.
There is two reasons:
- There was a timing issue in React 18 caused by the SelectControl that was harder to fix in the linked lists because of the way it handles async search.
- We are replacing the usage of select control in other places as well, so this stays in sync with that.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Load this branch and enable the new product editor in **WooCommerce > Settings > Advanced > Features**
2. Go to **Products > Add New**
3. Create a couple products with categories / tags
4. Once you created a couple products go to **Products > Add New** again
5. Add a title a category or tag ( similar to your previous products )
6. Go to **Linked products** tab
7. Add a couple products by using the search functionality in both Upsells and Cross-sells
8. Once you select an item the selected item should not show up in the dropdown anymore
9. If you save the product and search, the product your editing should not show up in the dropdown either.
10. Click **Choose products for me** this should add products that match by tag or category ( similar as before )
11. Save the product
12. Refresh the page the selected products should load/render correctly.
13. Go to the **General** tab and refresh with your network panel open
14. Make sure no request that starts with `wc/v3/products?search` is done
15. Now navigate to the **Linked products** tab
16. This should trigger a `wc/v3/products?search` request
17. Make sure the Linked E2E tests work correctly, you can run them locally with: `pnpm test:e2e-pw ./tests/e2e-pw/tests/merchant/products/block-editor/linked-product-tab-product-block-editor.spec.js` ( see E2E readme for setup: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/tests/e2e-pw/README.md )
18. Now merge `fix/wcadmin-react-18-create-root` branch into this one and build it
19. Run the above E2E tests again, they should pass successfully ( previously these would fail ).

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

louwie17 committed:
Make use of GB Combobox for the product select

louwie17 committed:
Update E2E tests

louwie17 committed:
Add changelog

louwie17 commented on packages/js/product-editor/src/blocks/generic/linked-product-list/edit.tsx:
```diff
@@ -73,68 +74,98 @@ export function LinkedProductListBlockEdit( {
+			);
+			if ( isAlreadySelected ) {
+				return;
+			}
```
Added the duplicate check

louwie17 commented on packages/js/product-editor/src/blocks/generic/linked-product-list/edit.tsx:
```diff
@@ -73,68 +74,98 @@ export function LinkedProductListBlockEdit( {
+			);
+
+			setLinkedProductIds( newLinkedProductIds );
+			searchProducts( '', newLinkedProductIds );
```
Reset search results upon selection

louwie17 commented on packages/js/product-editor/src/blocks/generic/linked-product-list/reducer.ts:
```diff
@@ -89,33 +86,6 @@ export function getLoadLinkedProductsDispatcher(
-					},
-				} );
-				return response;
-			} );
```
This is partially moved to the `linked-product-list` edit component

louwie17 commented on packages/js/product-editor/src/components/product-select/product-select.tsx:
```diff
@@ -1,93 +1,207 @@
-	__experimentalSelectControlMenuItem as MenuItem,
-	useAsyncFilter,
+	BaseControl,
+	ComboboxControl as CoreComboboxControl,
```
I replaced the `__experimentalSelectControl` with the core `ComboboxControl`

louwie17 committed:
Small CSS changes

louwie17 commented on packages/js/product-editor/src/components/product-select/types.ts:
```diff
@@ -6,8 +6,19 @@ import { Product } from '@woocommerce/data';
+	isLoading?: boolean;
+	onAddNew?: ( value: string ) => void;
+	filter( value?: string ): Promise< Product[] > | void;
 	onSelect?( product: Product ): void;
```
Made sure the new props are optional, incase plugins where already using the `ProductSelect` component. Otherwise it should work the same as before.

louwie17 committed:
Fix create variable product E2E test

louwie17 committed:
Add changelogs

louwie17 committed:
Use our internal wrapper of the combobox

louwie17 committed:
A couple small fixes

octaedro reviewed (commented):
While testing this PR I noticed a couple of bugs.
After finishing step 9, I refreshed the page and the product I was creating was added to both lists:

https://github.com/user-attachments/assets/caa52e50-2666-4e7b-a771-c88a4c5658a7

After that, I pressed `Choose products for me`, and it cleaned the lists I created. In the video, I removed the item I was creating before pressing the button, but it will clear the list with that item too.

octaedro commented:
I have a question regarding step 10 (_Click **Choose products for me** this should add products that match by tag or category ( similar as before_). Is it expected that the button will delete the items I already selected? I'm not sure it should behave that way.

https://github.com/user-attachments/assets/3b437eb8-5d87-47dc-b61f-26ce7ea401c8

I would assume that the user would want to keep the items they manually selected before.

louwie17 committed:
Fix bug and address feedback

louwie17 commented:
@octaedro I fixed the bug you found on the current product getting included, @mdperez86 had pointed that out to, I was using `.push` instead of using a spread operator.

> I have a question regarding step 10 (Click Choose products for me this should add products that match by tag or category ( similar as before). Is it expected that the button will delete the items I already selected? I'm not sure it should behave that way.

This is expected, both the previous logic was written in that manner and the E2E test also expects this. I actually didn't change any of the logic around the `Choose products for me` I left that the same.

mdperez86 reviewed (commented):
This is looking good. Tested according the description of the issue and the linked product search is working as expected.

I'm pre-approving this. But I would like @octaedro to give his final review!
Nice job @louwie17 !!!

octaedro commented on packages/js/product-editor/src/blocks/generic/linked-product-list/edit.tsx:
```diff
@@ -68,73 +70,125 @@ export function EmptyStateImage( {
-		if ( ! state.selectedProduct ) {
+		if (
+			! state.selectedProduct &&
+			linkedProductIds &&
```
With this check here, we won't need the nullish coalescing operator a couple of lines below
```
loadLinkedProductsDispatcher( linkedProductIds ?? [] );
```

octaedro commented on packages/js/product-editor/src/blocks/generic/linked-product-list/edit.tsx:
```diff
@@ -68,73 +70,125 @@ export function EmptyStateImage( {
+	}, 300 );
 
 	useEffect( () => {
 		// Only filter when the tab is selected.
```
I think that this comment should be updated. We can say:

```
// Only filter when the tab is selected and initial search results haven't been loaded yet.
```

octaedro commented on packages/js/product-editor/src/blocks/generic/linked-product-list/edit.tsx:
```diff
@@ -143,6 +197,15 @@ export function LinkedProductListBlockEdit( {
 		);
 
 		setLinkedProductIds( newLinkedProductIds );
+		setSearchValue( '' );
+		setIsSearching( true );
+		searchProducts( '', [ ...( linkedProductIds || [] ), productId ] )
+			.then( ( products ) => {
+				setSearchedProducts( products );
+			} )
+			.finally( () => {
+				setIsSearching( false );
+			} );
```
We repeated these lines a few times. It would be great to have an abstraction here.

louwie17 replied:
I originally had this abstracted and Maikel had suggested moving it out of the component, but given its predominantly state changes it did cause this duplicate code.
I end up moving it to its own function again: https://github.com/woocommerce/woocommerce/pull/50128/commits/b3b851835779ae238ed537748a356229bfc4ec24

octaedro commented on packages/js/product-editor/src/components/product-select/product-select.tsx:
```diff
@@ -1,93 +1,173 @@
+ * Map the product item to the Combobox core option.
+ *
+ * @param {Product} attr - Product item.
+ * @return {ComboboxControlOption}               Combobox option.
```
Is there a reason for the extra spaces?

louwie17 replied:
https://github.com/woocommerce/woocommerce/pull/50128/commits/15b233568a699f7bba91f98801af09c93207ef4f

octaedro commented on packages/js/product-editor/src/components/product-select/product-select.tsx:
```diff
@@ -1,93 +1,173 @@
+/**
+ * ComboboxControlOption component.
+ *
+ * @return {JSX.Element}                       Component item.
```
Same here

louwie17 replied:
https://github.com/woocommerce/woocommerce/pull/50128/commits/15b233568a699f7bba91f98801af09c93207ef4f

octaedro reviewed (commented):
Good job @louwie17! This is testing well on my end and the code looks good to me. I just left a couple of suggestions, nothing big.

louwie17 committed:
Address small formatting in comments

louwie17 committed:
Move search products back to consolidated function

louwie17 commented:
> Good job @louwie17! This is testing well on my end and the code looks good to me. I just left a couple of suggestions, nothing big.

I addressed your comments, thanks!
Btw I don't know if we really have a rule for this, and is probably just a preference or pet peeve of mine, but on your initial review it would of been nice to already hear about the formatting issues you mentioned in this second review as those were already present than to avoid the extra back and forth.
Or explicitly mention that you **only** tested the PR found the bugs and will do a deeper dive of the code once the bugs are fixed.

octaedro reviewed (approved):
Thank you @louwie17, for addressing the changes. This is testing well here.
Good job! 🚀
